File: Алгоритмы и структуры данных/Tim sort/stack.cpp
#define stack_var int

struct node
{
    stack_var start_index = 0;
    stack_var length = 0;
    node *next;
};

class stack
{
private:

public:
    node null_node;
    node *head;
    int size = 0;

    stack()
    {
        head = nullptr;
    }

    bool is_empty()
    {
        return head == nullptr;
    }

    void push(stack_var run_start_index, stack_var run_length)
    {
        node *tmp = new node;
        tmp->start_index = run_start_index;
        tmp->length = run_length;
        tmp->next = head;
        head = tmp;
        size++;
    }

    void pop()
    {
        if (!is_empty())
        {
            node *tmp = head;
            head = head->next;
            size--;
            delete tmp;
        }
    }

    node peek(int index = 0)
    {
        if (!is_empty() && index < size)
        {
            node *tmp = head;
            for (int i = 0; i < index && index < size; i++)
            {
                tmp = tmp->next;
            }

            return *tmp;
        }
        else
        {
            return null_node;
        }
    }

    void edit(stack_var run_start_index, stack_var run_length, int index = 0)
    {
        if (!is_empty() /*&& index < size*/)
        {
            node *tmp = head;
            /* Лишнии проверки в данном конкретном случае
            for (int i = 0; i < index && index < size; i++)
            {
                tmp = tmp->next;
            }
            */
            tmp->start_index = run_start_index;
            tmp->length = run_length;
        }
    }

    ~stack()
    {
        while (!is_empty())
        {
            pop();
        }
    }
};

File: Алгоритмы и структуры данных/Краскал на Java/Stack.java
package org.example;

class Stack<T>
{
    private class Node
    {
        T values;
        Node next;

        Node(T values)
        {
            this.values = values;
        }
    }

    Node nullNode;
    Node head;
    int size;

    public Stack(T nullNode)
    {
        this.nullNode = new Node(nullNode);
        this.head = null;
        this.size = 0;
    }

    boolean isEmpty()
    {
        return head == null;
    }

    public void push(T values)
    {
        Node tmp = new Node(values);
        tmp.next = head;
        head = tmp;
        size++;
    }

    void pop()
    {
        if (!isEmpty())
        {
            Node tmp = head;
            head = head.next;
            size--;
        }
    }

    public T peek(int index)
    {
        if (!isEmpty() && index < size)
        {
            Node tmp = head;
            for (int i = 0; i < index; i++)
            {
                tmp = tmp.next;
            }

            return tmp.values;
        } else
        {
            return nullNode.values;
        }
    }

    public void edit(T values, int index)
    {
        if (!isEmpty() && index < size)
        {
            Node tmp = head;
            tmp.values = values;
        }
    }
}